pkg/cmd/provisioning/roles.go

```diff
@@ -0,0 +1,337 @@
+	})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get IAM Identity Provider")
+	}
+
+	return *idProvider.Url, nil
+}
+
+func iamRolesCmd(cmd *cobra.Command, args []string) {
+	cfg := &awssdk.Config{
+		Region: awssdk.String(CreateOpts.Region),
+	}
+
+	s, err := session.NewSession(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	awsClient := aws.NewClientFromSession(s)
+
+	err = createIAMRoles(awsClient, CreateOpts.IdentityProviderARN, CreateOpts.NamePrefix, CreateOpts.CredRequestDir, CreateOpts.TargetDir, CreateOpts.GenerateOnly)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// StatementEntry is a simple type used to serialize to AWS' PolicyDocument format.
+type StatementEntry struct {
+	Effect   string
+	Action   []string
+	Resource string
+	// Must "omitempty" otherwise we send unacceptable JSON to the AWS API when no
+	// condition is defined.
+	Condition credreqv1.IAMPolicyCondition `json:",omitempty"`
+}
+
+// PolicyDocument is a simple type used to serialize to AWS' PolicyDocument format.
+type PolicyDocument struct {
+	Version   string
+	Statement []StatementEntry
+}
+
+func createRolePolicy(statements []credreqv1.StatementEntry) string {
+	policyDocument := PolicyDocument{
+		Version:   "2012-10-17",
+		Statement: []StatementEntry{},
+	}
+
+	for _, entry := range statements {
+		policyDocument.Statement = append(policyDocument.Statement,
+			StatementEntry{
+				Effect:    entry.Effect,
+				Action:    entry.Action,
+				Resource:  entry.Resource,
+				Condition: entry.PolicyCondition,
+			})
+	}
+
+	b, err := json.Marshal(&policyDocument)
+	if err != nil {
+		log.Fatalf("Failed to marshal the policy to JSON: %s", err)
+	}
+
+	return string(b)
+}
+
+// NewIAMRolesSetup provides the "create iam-roles" subcommand
+func NewIAMRolesSetup() *cobra.Command {
+	iamRolesSetupCmd := &cobra.Command{
+		Use: "iam-roles",
+		Run: iamRolesCmd,
+	}
+
+	iamRolesSetupCmd.PersistentFlags().StringVar(&CreateOpts.NamePrefix, "name-prefix", "", "Name prefix for all created AWS resources")
+	iamRolesSetupCmd.MarkPersistentFlagRequired("name-prefix")
+	iamRolesSetupCmd.PersistentFlags().StringVar(&CreateOpts.Region, "region", "", "AWS region where the S3 OpenID Connect endpoint will be created")
+	iamRolesSetupCmd.MarkPersistentFlagRequired("region")
+	iamRolesSetupCmd.PersistentFlags().StringVar(&CreateOpts.CredRequestDir, "credentials-requests-dir", "", "Directory containing files of CredentialsRequests to create IAM Roles for")
+	iamRolesSetupCmd.MarkPersistentFlagRequired("credentials-requests-dir")
+	iamRolesSetupCmd.PersistentFlags().StringVar(&CreateOpts.IdentityProviderARN, "identity-provider-arn", "", "ARN of IAM Identity provider for IAM Role trust relationship")
+	iamRolesSetupCmd.MarkPersistentFlagRequired("identity-provider-arn")
+	iamRolesSetupCmd.PersistentFlags().BoolVar(&CreateOpts.GenerateOnly, "generate-files-only", false, "Skip creating objects, and just save what would have been created into files")
```

Comment:
WDYT about -o json here?

Reply:
If we did something like '-o json' it might get a little weird with a high-level 'ccoctl create all' command where not all of the output files will be of type JSON (the public/private key). And I was planning on having the installer manifests generation output yaml files (not for any particular reason other than they're easier to edit IMO). The reason I'm generating JSON for the aws CLI files, is that it seems like that is what the tool provides when you ask for a skeleton object (I would've preferred YAML to be honest) WDYT?

Reply:
Yup that's fair. I'd like a more concise CLI param though. What about --dry-run and the command arguments inform you it will write files?